### PR TITLE
feat(common-scripts): integrete omnilock p2pkh

### LINKED ISSUE #398: [feature request] Support OmniLock in common-script
## Motivation

OmniLock is a widely used lock on the CKB chain. It is better to support omniLock so that users can easily forge an omnilock transfer transaction.

```ts
// omniLock.ts
export type OmniLockFromInfo = {
  address?: string;
  pubkeyAndType?: {
    pubkey: string;
    type: "ETH" | "SECP256K1"
  }
}
export async function transfer(
  txSkeleton: TransactionSkeletonType,
  fromInfo: OmniLockFromInfo,
  toAddress: Address,
  capacity: BIish,
  ...
): Promise<TransactionSkeletonType> 
```

## Preview

```ts
const { omniLock } = require("@ckb-lumos/common-scripts")
const { sealTransaction } = require("@ckb-lumos/helpers")

let txSkeleton = TransactionSkeleton({ cellProvider: indexer })

txSkeleton = await omniLock.transfer(
  txSkeleton,
  fromInfos, // OmniLockFromInfo
  "ckb1qyqrdsefa43s6m882pcj53m4gdnj4k440axqdt9rtd",
  BigInt(3500 * 10 ** 8),
)

// Signing messages will fill in `txSkeleton.signingEntries`.
txSkeleton = await omniLock.prepareSigningEntries(
  txSkeleton
)

// Then you can sign messages in order and get signatures.
// Call `sealTransaction` to get a transaction.
const tx = sealTransaction(txSkeleton, signatures)
```

## Packages Affected

- packages/common-scripts

### PR DESCRIPTION
**Pull request by zhangyouxin**
# Description

[Omnilock](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0042-omnilock/0042-omnilock.md) has become a very commonly used lock on CKB chain. This PR supports two most frequently used Omnilock features: `ETHEREUM` mode and `SECP256K1_BLAKE160` mode, so that users can transfer CKBs from an omnilock very easily.

## Usage:
 
```ts
import { common, omnilock } from '@ckb-lumos/common';

let txSkeleton = TransactionSkeleton({ cellProvider: indexer });
const fromLock: Script = omnilock.createOmnilockScript({ 
  auth: { 
    flag: 'ETHEREUM', // or 'SECP256K1_BLAKE160'
    content: '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
  }
})

const fromAddress = helpers.encodeToAddress(fromLock);
// another address 
const recieverAddress = "ckb1qyqrdsefa43s6m882pcj53m4gdnj4k440axqdt9rtd";

// transfer 10000 CKB 
txSkeleton = await common.transfer(
  txSkeleton, 
  [fromAddress],
  recieverAddress,
  10000e8
)

txSkeleton = common.prepareSigningEntries(txSkeleton)
const message = txSkeleton.get("signingEntries").get(0)!.message;
// use your own sign method
const sig = personalSign(message, ALICE_PRIVKEY);
const omnilockSig = bytes.hexify(
  omnilock.OmnilockWitnessLock.pack({ signature: sig })
);
const tx = sealTransaction(txSkeleton, [omnilockSig]);
const hash = await rpc.sendTransaction(tx, "passthrough");
```

Fixes #398 (issue)

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] [debugger test](https://github.com/zhangyouxin/lumos/blob/omnilock-common-scripts/packages/common-scripts/tests/omnilock.test.ts)
- [x] [example](https://github.com/zhangyouxin/lumos/blob/omnilock-common-scripts/packages/common-scripts/examples/omnilock/common.ts) (use ts-node to run it on testnet)

### CODE REVIEW

**Comment by homura** [comment]
please add test to pass the coverage checking

File: examples/omni-lock-secp256k1-blake160/lib.ts [hunk 1]
```diff
@@ -35,85 +21,26 @@ export interface TransferOptions {
...
+    txSkeleton,
+    [fromAddress],
+    toAddress,
+    BigInt(70*10**8),
+    undefined,
+    undefined,
+    {config: CONFIG}
+  )
+  
+  txSkeleton = await commons.common.payFee(txSkeleton,  [fromAddress], 1000, undefined, {config: CONFIG})
```
Comment (homura) [inline examples/omni-lock-secp256k1-blake160/lib.ts]: we can remove toMessage and replace with `prepareSigningEntries` here

File: examples/omni-lock-secp256k1-blake160/lib.ts [hunk 2]
```diff
@@ -139,9 +66,9 @@ export async function signByPrivateKey(txSkeleton: helpers.TransactionSkeletonTy
   const signature = hd.key.signRecoverable(messages.message, privateKey);
 
   const signedWitness = bytes.hexify(blockchain.WitnessArgs.pack({
-    lock: bytes.hexify(SerializeRcLockWitnessLock({
-      signature: bytes.bytify(signature),
-    })),
+    lock: commons.omnilock.OmnilockWitnessLock.pack({
```
Comment (homura) [inline examples/omni-lock-secp256k1-blake160/lib.ts]: is it necessary here? can we refactor the `toMessage` with `prepareSigningEntries`?

File: packages/common-scripts/tests/omnilock.test.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+const ckbConfig: Config = {
+  PREFIX: "ckt",
+  SCRIPTS: context.scriptConfigs,
+};
+
+test.before(async () => {
+  if (process.env.CKB_DEBUGGER_PATH) return;
+  await downloader.downloadIfNotExists();
+});
+// TODO uncomment the skip when ci is ready
```
Comment (homura) [inline packages/common-scripts/tests/omnilock.test.ts]:
```suggestion

```

File: packages/common-scripts/src/omnilock.ts [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+export const OmnilockWitnessLock = table(
+  {
+    signature: BytesOpt,
+    omni_identity: OmniIdentityOpt,
+    preimage: BytesOpt,
+  },
+  ["signature", "omni_identity", "preimage"]
+);
+
+export const CellCollector: CellCollectorConstructor = class CellCollector
```
Comment (homura) [inline packages/common-scripts/src/omnilock.ts:118]: why we need a `CellCollector` here?

File: packages/common-scripts/src/omnilock.ts [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  const omnilockConfig = config.SCRIPTS.OMNILOCK;
+  if (!omnilockConfig) {
+    throw new Error("OMNILOCK script config not found.");
+  }
+
+  // omni flag       pubkey hash   omni lock flags
+  // chain identity   eth addr      function flag()
+  // 00: Nervos       👇            00: owner
+  // 01: Ethereum     👇            01: administrator
+  let args;
```
Comment (PainterPuppets) [inline packages/common-scripts/src/omnilock.ts]: suggestion: maybe it is better avoids the use of `let` and `else`

``` ts 
const args = (() => {
  if (omnilockInfo.auth.flag === "ETHEREUM") {
    return `0x01${bytes.hexify(omnilockInfo.auth.content).slice(2)}00`;
  }

  if (omnilockInfo.auth.flag === "SECP256K1_BLAKE160") {
    return `0x00${bytes.hexify(omnilockInfo.auth.content).slice(2)}00`;
  }

  throw new Error(`Not supported flag: ${omnilockInfo.auth.flag}.`);
})()
```

File: packages/common-scripts/tests/helper.test.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import { isOmnilockAddress } from "../src/helper";
+import { Script } from "@ckb-lumos/base";
+import { encodeToAddress } from "@ckb-lumos/helpers";
+const { AGGRON4 } = predefined;
+
+test("should isOmnilockAddress return expected", (t) => {
+  const ALICE_PRIVKEY =
+    "0x1234567812345678123456781234567812345678123456781234567812345678";
+  const aliceArgs = hd.key.privateKeyToBlake160(ALICE_PRIVKEY);
+  console.log("alice args is:", aliceArgs);
```
Comment (PainterPuppets) [inline packages/common-scripts/tests/helper.test.ts]: remove console

File: packages/common-scripts/tests/helper.test.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import test from "ava";
+import * as hd from "@ckb-lumos/hd";
+import { omnilock } from "../src";
+import { predefined } from "@ckb-lumos/config-manager";
+import { isOmnilockAddress } from "../src/helper";
+import { Script } from "@ckb-lumos/base";
+import { encodeToAddress } from "@ckb-lumos/helpers";
+const { AGGRON4 } = predefined;
+
+test("should isOmnilockAddress return expected", (t) => {
```
Comment (PainterPuppets) [inline packages/common-scripts/tests/helper.test.ts]: need the `should other lock return false` testcase

File: packages/common-scripts/tests/omnilock.test.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+test("omnilock#common transfer", async (t) => {
+  const ALICE_PRIVKEY =
+    "0x1234567812345678123456781234567812345678123456781234567812345678";
+  const aliceArgs = hd.key.privateKeyToBlake160(ALICE_PRIVKEY);
+  console.log("alice args is:", aliceArgs);
```
Comment (PainterPuppets) [inline packages/common-scripts/tests/omnilock.test.ts]: remove console

**Review by PainterPuppets — COMMENTED** [review]
LGTM

File: packages/common-scripts/src/omnilock.ts (hunk 4 above)
Comment (zhangyouxin) [reply packages/common-scripts/src/omnilock.ts:118]: CellCollector is required: https://github.com/ckb-js/lumos/blob/develop/packages/common-scripts/src/common.ts#L48

File: examples/omni-lock-secp256k1-blake160/lib.ts (hunk 2 above)
Comment (homura) [reply examples/omni-lock-secp256k1-blake160/lib.ts]: why it is marked as resolved?

**Review by PainterPuppets — APPROVED** [review]
LGTM

**Review by homura — APPROVED** [review]
LGTM

**Pull request merged**